cql3/statements/select_statement.cc

```diff
@@ -1499,70 +1502,132 @@ void select_statement::verify_ordering_is_allowed(const restrictions::statement_
+    if (orderings.empty()) {
+        return false;
+    }
+
+    return are_column_select_results_reversed(*orderings[0].first, orderings[0].second);
+}
+
+void select_statement::verify_ordering_is_valid(const prepared_orderings_type& orderings,
+                                                const schema& schema,
+                                                const restrictions::statement_restrictions& restrictions) const {
+    if (orderings.empty()) {
         return;
     }
 
-    for (auto&& def : schema.partition_key_columns()) {
-        if (!selection.has_column(def)) {
-            throw exceptions::invalid_request_exception(format("SELECT DISTINCT queries must request all the partition key columns (missing {})", def.name_as_text()));
+    // Verify that columns are in the same order as in schema definition
+    for (size_t i = 0; i + 1 < orderings.size(); i++) {
+        if (orderings[i].first->component_index() >= orderings[i + 1].first->component_index()) {
+            throw exceptions::invalid_request_exception("Order by currently only supports the ordering of columns following their declared order in the PRIMARY KEY");
         }
     }
-}
 
-void select_statement::handle_unrecognized_ordering_column(const column_identifier& column) const
-{
-    if (contains_alias(column)) {
-        throw exceptions::invalid_request_exception(format("Aliases are not allowed in order by clause ('{}')", column));
+    // We only allow leaving select results unchanged or reversing them. Find out if they should be reversed.
+    bool is_reversed = is_ordering_reversed(orderings);
+
+    // Now verify that
+    // 1. Each column in the specified clustering prefix either has an ordering, or an EQ restriction.
+    //    When there is an EQ restriction ordering is not needed because only a single value is possible.
+    // 2. The ordering leaves the selected rows unchanged or reverses the normal result.
+    //    For that to happen the ordering must reverse all columns or none at all.
+    uint32_t max_ck_index = orderings.back().first->component_index();
+    auto orderings_iterator = orderings.begin();
+    for (uint32_t ck_column_index = 0; ck_column_index <= max_ck_index; ck_column_index++) {
+        if (orderings_iterator->first->component_index() != ck_column_index) {
+            // We don't have ordering for this column. Fail with an error.
+            throw exceptions::invalid_request_exception("Order by currently only supports the ordering of columns following their declared order in the PRIMARY KEY");
+        }
+
+        // We have ordering for this column, let's see if its reversing is right.
+        const column_definition* cur_ck_column = orderings_iterator->first;
+        bool is_cur_column_reversed = are_column_select_results_reversed(*cur_ck_column, orderings_iterator->second);
+        if (is_cur_column_reversed != is_reversed) {
+            throw exceptions::invalid_request_exception("Unsupported order by relation");
```

Comment:
Perhaps we could print a more useful error message here - perhaps saying that either all or none of the clustering column orders can be reversed.

Reply:
:+1: changed to something better. 
I left it this way because I was worried that tests will break, but they actually don't check this message.

Reply:
Again, I think I don't see the new version here?